### PR TITLE
Add Resources to Recordable interface.

### PR DESCRIPTION
**Pull request by lalitb**
This is second part of initial Resource SDK implementation ( https://github.com/open-telemetry/opentelemetry-cpp/pull/502 ) - to propagate resources to Span exporters through Recordable interface.

TODO -> Modify otlp exporter to export resources. This is dependent on #567 , as the spans need to be grouped on basis on Resources and Instrumentation Library before exporting.

@jsuereth  - It doesn't use alternate approach from processor to exporters, as that would need re-work if Resources are later made as mutable / append-only attributes ( as there are some discussion happening on this (and you are already part of that disucssions :) ) : https://github.com/open-telemetry/opentelemetry-specification/issues/1298 ).

Moving to draft, as per comments from Josh.

### CODE REVIEW

File: exporters/otlp/src/otlp_exporter.cc [hunk 1]
```diff
@@ -23,10 +23,23 @@ void PopulateRequest(const nostd::span<std::unique_ptr<sdk::trace::Recordable>>
...
   {
     auto rec = std::unique_ptr<Recordable>(static_cast<Recordable *>(recordable.release()));
     *instrumentation_lib->add_spans() = std::move(rec->span());
+
+    if (!resource_added)
+    {
+      auto resource = rec->GetResources();
+      if (resource.attributes_size())
+      {
+        // TODO
```
Comment (jsuereth) [inline exporters/otlp/src/otlp_exporter.cc:38]: Is this ready for review?

File: exporters/otlp/src/recordable.cc [hunk 2]
```diff
@@ -117,6 +117,101 @@ void PopulateAttribute(opentelemetry::proto::common::v1::KeyValue *attribute,
...
+                               const opentelemetry::sdk::common::OwnedAttributeValue &value)
+{
+
+  attribute->set_key(key.data(), key.size());
+
+  if (nostd::holds_alternative<bool>(value))
+  {
+    attribute->mutable_value()->set_bool_value(nostd::get<bool>(value));
+  }
+
```
Comment (ThomsonTan) [inline exporters/otlp/src/recordable.cc:131]: nit: remove blank line.

File: sdk/include/opentelemetry/sdk/trace/span_data.h [hunk 3]
```diff
@@ -121,6 +121,16 @@ class SpanData final : public Recordable
    */
   opentelemetry::trace::SpanKind GetSpanKind() const noexcept { return span_kind_; }
 
+  /**
+   * Get Resources.
+   * @return the resource of this span
+   */
+  const std::unordered_map<std::string, opentelemetry::sdk::common::OwnedAttributeValue>
+      &GetResources() const noexcept
```
Comment (jsuereth) [inline sdk/include/opentelemetry/sdk/trace/span_data.h:129]: Why doesn't this return a `Resource` object?

File: sdk/include/opentelemetry/sdk/trace/span_data.h [hunk 4]
```diff
@@ -213,6 +223,14 @@ class SpanData final : public Recordable
     span_kind_ = span_kind;
   }
 
+  void SetResourceAttribute(
```
Comment (jsuereth) [inline sdk/include/opentelemetry/sdk/trace/span_data.h:226]: I'm trying to understand what this method is intended to do and where it is in the spec.

Fro what I gather, this is basically copying all the resource-attributes onto the span.  Can we instead just grab a reference to the resource?

**Review by jsuereth — COMMENTED** [review]
IIUC, rather than passing resource to exporter, we're copying all the labels across spans and assuming they all share the same set?

This code was a bit harder to follow.    WDYT about one of the following:

1. Keep the `Resource` bundled into that type instead of using the name "resource attribtues"
2. Modify `MakeRecordable` interfaces to take in a resource
3. I'm happy to take a stab at an interface for sending resource from the TracerProvider.  (Specifically I'd add a `setResource` on `Processor` that is called when a processor is added to the `TracerProvider`).

**Comment by lalitb** [comment]
> IIUC, rather than passing resource to exporter, we're copying all the labels across spans and assuming they all share the same set?
> 
> This code was a bit harder to follow. WDYT about one of the following:
> 
> 1. Keep the `Resource` bundled into that type instead of using the name "resource attribtues"
> 2. Modify `MakeRecordable` interfaces to take in a resource
> 3. I'm happy to take a stab at an interface for sending resource from the TracerProvider.  (Specifically I'd add a `setResource` on `Processor` that is called when a processor is added to the `TracerProvider`).

Thanks @jsuereth for your comments.  I would prefer option 3 unless Resource is made append-only/mutable in spec. Will go with option 2 in that case. Would wait for more suggestions before doing changes accordingly.

**Comment by jsuereth** [comment]
@lalitb Sounds like a plan.  I'll get something to you as soon as I can (this week, likely)

**Comment by jsuereth** [comment]
Ok #575 is ready for comment/comparison.

**Comment by lalitb** [comment]
closing as we are looking into alternate options of implementation ref: #580 , #591

**Pull request closed without merging**